storage/innobase/row/row0merge.cc

```diff
@@ -2286,6 +2303,48 @@ row_merge_read_clustered_index(
 			}
 		}
 
+		if (old_table->versioned()) {
+			if (new_table->versioned() && !drop_historical) {
+				dfield_t *end = dtuple_get_nth_field(
+					row, new_table->vers_end);
+				byte *data = static_cast<byte *>(
+					dfield_get_data(end));
+				ut_ad(data);
+				if (mach_read_from_8(data) == TRX_ID_MAX) {
+					dfield_t *start = dtuple_get_nth_field(
+						row, new_table->vers_start);
+					void *data = dfield_get_data(start);
+					ut_ad(data);
+					mach_write_to_8(data, trx->id);
+					trx->vers_update_trt= true;
+				}
+			} else {
+				const dict_col_t *col =
+					&old_table->cols
+						 [old_table->vers_end];
+				const ulint nfield = dict_col_get_clust_pos(
+					col, clust_index);
+				ulint len = 0;
+				const rec_t *sys_trx_end = rec_get_nth_field(
+					rec, offsets, nfield, &len);
+				ut_ad(len == 8);
+				if (mach_read_from_8(sys_trx_end) != TRX_ID_MAX)
+					continue;
+			}
+		} else if (new_table->versioned()) {
+			void *sys_trx_start = mem_heap_alloc(row_heap, 8);
+			void *sys_trx_end = mem_heap_alloc(row_heap, 8);
+			mach_write_to_8(sys_trx_start, trx->id);
+			mach_write_to_8(sys_trx_end, TRX_ID_MAX);
```

Comment:
Please use stack-local variables for these, and initialize them only once, not for every row.